### PR TITLE
Implement the WAGED rebalance with delayed rebalance logic.

### PR DESCRIPTION
**Pull request by jiajunwang**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR title:

#402

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This PR implements the WAGED rebalancer.
Delayed rebalance logic will be integrated with the WAGED rebalancer later.

### Tests

- [ ] The following tests are written for this issue:

TBD

- [ ] The following is the result of the "mvn test" command on the appropriate module:

TBD

### Commits

- [ ] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation in the following wiki page:

https://github.com/apache/helix/wiki/Weight-aware-Globally-Evenly-distributed-Rebalancer

### Code Quality

- [ ] My diff has been formatted using helix-style.xml

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java [hunk 1]
```diff
@@ -41,18 +60,279 @@
 public class WagedRebalancer implements GlobalRebalancer<ResourceControllerDataProvider> {
   private static final Logger LOG = LoggerFactory.getLogger(WagedRebalancer.class);
 
+  // When any of the following change happens, the rebalancer needs to do a global rebalance.
+  private static final Set<HelixConstants.ChangeType> GLOBAL_REBALANCE_REQUIRED_CHANGE_TYPES =
+      Collections.unmodifiableSet(new HashSet<>(Arrays
```
Comment (i3wangyi) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java:62]: Use ImmutableSet.of

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java [hunk 2]
```diff
@@ -41,18 +60,279 @@
...
+          .asList(HelixConstants.ChangeType.RESOURCE_CONFIG, HelixConstants.ChangeType.CONFIG,
+              HelixConstants.ChangeType.INSTANCE_CONFIG)));
+
+  private ResourceChangeDetector _clusterDataDetector;
+  private AssignmentMetadataStore _assignmentMetadataStore;
+  private RebalanceAlgorithm _rebalanceAlgorithm;
+  private MappingCalculator<ResourceControllerDataProvider> _mappingCalculator;
+
   @Override
   public void init() {
```
Comment (i3wangyi) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java]: I don't think the init method is necessary. If the class is stateless we can create singleton instance. A constructor will give you the ability to mock the dependency for unit testing

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java [hunk 3]
```diff
@@ -41,18 +60,279 @@
...
+   *                           Note the resources that are not configured to use the WAGED rebalancer will be ignored.
+   * @param currentStateOutput The processed current states information.
+   * @return A map of rebalanced IdealState.
+   * @throws HelixException if the rebalance fails unexpectedly.
+   */
   @Override
-  public Map<String, IdealState> computeNewIdealStates(ResourceControllerDataProvider clusterData,
-      Map<String, Resource> resourceMap, final CurrentStateOutput currentStateOutput) {
-    return new HashMap<>();
+  public Map<String, IdealState> computeNewIdealStates(ResourceControllerDataProvider dataProvider,
```
Comment (i3wangyi) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java]: The public method and the related private methods are too long and unbelievably difficult to understand and maintain

Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java]: I have divided the PR to 2 pieces. But in general, this logic is one of the most complicated ones in our project. That's why we need to start early.
And, believe me, split them into sub methods won't help with readability...

Please try to understand what we are doing and comment with detail if you think any part can be simplified.

**Pull request closed without merging**